### PR TITLE
Add softmax example

### PR DESCRIPTION
**Pull request by Schweini-PEK**
## Add softmax example — multi-case optimized version

Online safe softmax with multi-kernel strategy for arbitrary dim, dtype, and shape. This is a **performance-optimized version targeting cann-bench multi-case evaluation**, coexisting with the existing `example_online_softmax.py` (simple single-kernel demo) — **does not overwrite it**.

### Features

- Supports float16, float32, bfloat16
- Arbitrary dim (0 to rank-1, negative index) via 4 fast paths:
  - **Fast path 1**: 2D dim=last (direct kernel call)
  - **Fast path 2**: N-D dim=last (zero-copy reshape, no permute)
  - **Fast path 3**: 2D dim=0 (specialized `online_softmax_dim0` kernel, reduce along M)
  - **Fast path 4**: N-D interior dim (3D `online_softmax_3d` kernel, reduce along B in [A,B,C])
- Non-aligned N handled via `pad_value=-inf` in T.copy (required for softmax correctness — backend auto tail-block fill uses 0, which corrupts the reduce_sum denominator)
- fp16/bf16 use float32 compute internally (meets cann-bench MERE/MARE thresholds)
- Simplified single-pass kernel for n_num=1 (small N, e.g. N=2)

### Optimizations

- Online safe softmax two-pass algorithm (Pass 1: max+sum, Pass 2: normalize)
- Pass 2 Double Buffer with manual MTE2/V/MTE3 three-stage flag pipeline (AUTO_SYNC=False) — fp16/bf16 only
- Dual-kernel strategy: use_db=True uses Double Buffer; use_db=False uses AUTO_SYNC=True (avoids barrier overhead)
- 3D interior-dim kernel eliminates `.permute().contiguous()` overhead for dim≠0,≠last
- dim=0 specialized kernel eliminates permute for dim=0 cases
- Instruction fusion: `mul+add` → `mul_add_dst` (guide §2.7)
- Adaptive block_M/block_N with UB budget formula (guide §2.11)
- Parallelism check: reduce block_M to 16 when m_num < core_num (guide §2.9 + antipatterns §B)

### About AUTO_SYNC=False

Pass 2 Double Buffer uses `AUTO_SYNC=False` with manual `set_flag`/`wait_flag` for MTE2→V and V→MTE3 dependencies. The three-stage pipeline works as follows:

1. **Prefetch**: Load tile 0 into stage 0 buffer → `set_flag("mte2", "v", 0)`
2. **Main loop** (per tile): 
   - Prefetch next tile into `nxt` stage → `set_flag("mte2", "v", nxt)`
   - Consume current `cur` stage: cast → sub → exp → div → cast → `set_flag("v", "mte3", cur)`
   - Store current: `wait_flag("v", "mte3", cur)` → copy to GM → `set_flag("mte3", "mte2", cur)` (release buffer)
3. **Epilogue**: Consume and store last prefetched tile

Flag chain: `mte3→mte2` (buffer released) → `mte2→v` (input ready) → `v→mte3` (output ready)

Pass 1 (which has online cross-iteration dependencies) uses `AUTO_SYNC=True` via the dual-kernel strategy — Pass 1 DB was attempted but caused 507014 AICore errors due to flag chain conflicts with in-place online update operations.

### Relationship with existing `example_online_softmax.py`

| File | Lines | Description |
|------|-------|-------------|
| `examples/softmax/example_online_softmax.py` | 130 | Original simple single-kernel demo (unchanged) |
| `examples/softmax-new/example_softmax.py` | 1375 | This PR — multi-case optimized version with wrapper, golden, cann-bench 20 cases, L0/L1/L2/Boundary test suite |

The original `examples/softmax/` is **not modified**. This PR adds `examples/softmax-new/example_softmax.py` as a separate file with full wrapper (arbitrary dim/dtype/shape), PyTorch golden, cann-bench precision checker, and layered test suite.

### Backend: AscendC (not PTO)

This PR uses the **AscendC backend** (`target="auto"` → `llvm --keys=ascend`). The PTO backend (`target="pto"`) was tested locally but encountered a compilation segfault in the PTO codegen path (`tilelang_ascend_pto`). The `bisheng` compiler is available, but the PTO codegen pass crashes. This is left as a future improvement — switching to PTO requires adding `target="pto"` to all 5 `@tilelang.jit` decorators and resolving the codegen issue.

### pad_value=-inf note

The tail-block `pad_value=-inf` in `T.copy` is **required** for softmax correctness. We tested removing it (relying on backend auto tail-block handling), but the backend fills tail blocks with **0** instead of **-inf**. Since `exp(0)=1` gets added to the `reduce_sum` denominator, the output is systematically underestimated (e.g. `y_sum=0.903` instead of `1.0` for case 8 `[2049,4097]`). This causes precision failures on all non-aligned-N cases. The `pad_value=-inf` must be kept.

### Performance Anti-pattern Self-check

| Anti-pattern | Status | Note |
|---|---|---|
| launch core count too high (A) | Tested, not adopted | Hardware scheduler handles multi-block efficiently |
| Vector per-row for loop | Not present | All operations use tile-level broadcast + vectorized instructions |
| Redundant global sync | Minimized | Pass 2 uses flag sync; Pass 1 uses barrier_all (AUTO_SYNC=True path) |
| Unfused instructions | Optimized | `mul+add` fused to `mul_add_dst` |
| Tile size too small | Optimized | UB budget formula dynamically computes max block_N (up to 512) |
| AIC/AIV CV overlap | N/A | Pure AIV operator |
| AIV memory bound without pipeline | Optimized | Pass 2 Double Buffer with three-stage pipeline |

### Operator Evaluation Report

**Latest local cann-bench (CANN 9.0.0, Ascend910B3, AscendC backend)**:

| Metric | Value |
|--------|-------|
| pass_cases | 20/20 (100%) |
| avg_speedup | 0.64x |
| score | 67.90 |

**Per-case speedup (local)**:

| Case | Shape | dtype | dim | speedup |
|------|-------|-------|-----|---------|
| 1 | [1024,1024] | fp16 | -1 | 1.24x |
| 2 | [2048,2048] | fp32 | -1 | 0.80x |
| 3 | [4096,4096] | bf16 | -1 | 0.58x |
| 4 | [8192,8192] | fp16 | 0 | 1.16x |
| 5 | [8192,8192] | fp32 | 1 | 0.62x |
| 6 | [31,67,127,257] | bf16 | 2 | 0.38x |
| 7 | [1023,2047] | fp16 | -1 | 0.77x |
| 8 | [2049,4097] | fp32 | -1 | 0.47x |
| 9 | [127,257,1023] | bf16 | -2 | 0.54x |
| 10 | [1009,1021] | fp16 | -1 | 0.94x |
| 11 | [367,373,379] | fp32 | 1 | 0.75x |
| 12 | [11,13,17,4001] | bf16 | -1 | 0.64x |
| 13 | [1000003,2] | fp16 | -1 | 0.01x |
| 14 | [11,13,17,67,67] | fp32 | -1 | 0.41x |
| 15 | [3,7,11,13,1013] | bf16 | -1 | 0.87x |
| 16 | [512,2049] | fp16 | 0 | 0.49x |
| 17 | [255,8193] | fp32 | 1 | 0.30x |
| 18 | [2,511,2049] | bf16 | -1 | 0.69x |
| 19 | [4,255,2049] | fp16 | 1 | 0.64x |
| 20 | [2,3,17,1024,101] | fp32 | 3 | 0.55x |

**Online cann-bench (910C, CANN 9.0.0, AscendC backend)**: 20/20 pass, score 67.03

**Case 17 optimization**: `block_M` reduced from 128 to 16 when `m_num < core_num`, increasing parallelism from 2 cores to 16 cores. Online speedup: 0.29x → 0.45x (+52.9%).

Profiling:

softmax-prof

### Known limitations

- **Case 13** `[1000003,2] fp16` (speedup 0.01x): N=2 is extremely small (4 bytes per row), DMA 32B alignment wastes 87.5%. TileLang kernel cannot efficiently handle this. Pipeline optimization was attempted but caused aicore deadlock (flag initialization optimized away by compiler). Left as future work.
- **PTO backend**: Not supported (compilation segfault in codegen). Uses AscendC backend.
- **fp32 Double Buffer**: Tested but resulted in net negative performance — DB's block_N tightening (UB budget) increased n_num iterations, offsetting pipeline overlap gains. Reverted; fp32 uses AUTO_SYNC=True path.

### CODE REVIEW

File: examples/cann-bench/softmax/example_softmax.py [hunk 1]
```diff
@@ -0,0 +1,1375 @@
+"""Softmax operator implementation using TileLang-Ascend.
```
Comment (benyang0506) [inline examples/cann-bench/softmax/example_softmax.py:1]: A softmax example already exists at https://github.com/tile-ai/tilelang-ascend/tree/ascendc_pto/examples/softmax. If you want to extend it, please do so in this folder rather than creating a new softmax-new directory.

File: examples/softmax-new/example_softmax.py [hunk 2]
```diff
@@ -0,0 +1,1375 @@
+"""Softmax operator implementation using TileLang-Ascend.
```
Comment (MiaoHaifeng) [inline examples/softmax-new/example_softmax.py:1]: 文件放到cann-bench下
